### PR TITLE
ARROW-12790: [C++] Improve HadoopFileSystem conformance

### PR DESCRIPTION
**Pull request by pitrou**
* Ensure the HadoopFileSystem meets most requirements from the FileSystem API.
* Implement HadoopFileSystem::CopyFile.
* Enable generic filesystem tests for HadoopFileSystem.
* Add generic filesystem test for special characters.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit test-*hdfs*

**Comment by pitrou** [comment]
@github-actions crossbow submit test-conda-python-3.7-hdfs-2.9.2 test-conda-python-3.7-hdfs-3.21

**Comment by pitrou** [comment]
@github-actions crossbow submit test-conda-python-3.7-hdfs-3.2.1

File: cpp/src/arrow/filesystem/test_util.cc [hunk 1]
```diff
@@ -453,9 +453,11 @@ void GenericFileSystemTest::TestMoveDir(FileSystem* fs) {
...
 
   // Destination is a non-empty directory
-  ASSERT_RAISES(IOError, fs->Move("KL", "EF"));
-  AssertAllDirs(fs, {"EF", "KL", "KL/CD"});
-  AssertAllFiles(fs, {"EF/ghi", "KL/CD/def", "KL/abc"});
+  if (!allow_move_dir_over_non_empty_dir()) {
+    ASSERT_RAISES(IOError, fs->Move("KL", "EF"));
+    AssertAllDirs(fs, {"EF", "KL", "KL/CD"});
+    AssertAllFiles(fs, {"EF/ghi", "KL/CD/def", "KL/abc"});
+  }
```
Comment (lidavidm) [inline cpp/src/arrow/filesystem/test_util.cc:460]: Is it worth testing the other side of this? (Presumably it'd succeed and we'd have `EF/KL/{abc, CD/def}`?)

File: cpp/src/arrow/filesystem/hdfs.cc [hunk 2]
```diff
@@ -134,21 +140,39 @@ class HadoopFileSystem::Impl {
...
-      wd = wd_uri.path();
+    RETURN_NOT_OK(client_->GetWorkingDirectory(&wd));
+
+    if (!select.base_dir.empty() && select.base_dir.front() == '/') {
+      // base_dir is absolute, only keep the URI authority portion.
+      // As mentioned in StatSelector() above, the URI may contain unescaped
+      // special chars and therefore may not be a valid URI, so we parse by hand.
+      auto pos = wd.find("://");  // start of host:port portion
+      if (pos == std::string::npos) {
+        return Status::Invalid("xxx");
```
Comment (lidavidm) [inline cpp/src/arrow/filesystem/hdfs.cc:157]: Were these supposed to be real error messages? :slightly_smiling_face:

**Comment by pitrou** [comment]
Crossbow builds submitted at https://github.com/ursacomputing/crossbow/branches/all?query=build-364

File: cpp/src/arrow/filesystem/hdfs.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/hdfs.cc:157]: Well, of course, they're just terse :-D
I'll fix them tomorrow.

File: cpp/src/arrow/filesystem/test_util.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/test_util.cc:460]: Yes, will do.

**Pull request closed without merging**